tools/gen_ut_cmakelists.py

```diff
@@ -161,146 +162,317 @@ def process_name(name, curdir):
-                set_tests_properties(test_allreduce PROPERTIES  TIMEOUT "120" RUN_SERIAL 1)
-            endif()"
+            for c in self.NO_CMAKE_DIR_WARNING:
+                err_msg += "   " + c + "\n"
+            err_msg += """
+        This may cause the dist ports different with the old version.
+        If the directories are newly created or there is no CMakeLists.txt before, or ignore this error, you
+        must specify the directories using the args option --ignore-cmake-dirs/-i.
+        If you want to keep the dist ports of old tests unchanged, please ensure the old
+        verson CMakeLists.txt file existing before using the gen_ut_cmakelists tool to 
+        generate new CmakeLists.txt files.
+    ====================================================================================
     """
-
-    name, os_, archs, timeout, run_type, launcher, dist_ut_port, run_serial, envs, conditions = line.strip(
-    ).split(",")
-
-    # name == "name" means the line being parsed is the header of the table
-    # we should skip this line and return empty here.
-    if name == "name":
-        return ""
-    name = process_name(name, curdir)
-
-    envs = process_envs(envs)
-    conditions = process_conditions(conditions)
-    archs = proccess_archs(archs)
-    os_ = process_os(os_)
-    run_serial = process_run_serial(run_serial)
-    run_type = process_run_type(run_type)
-
-    cmd = ""
-
-    for c in conditions:
-        cmd += f"if ({c})\n"
-
-    time_out_str = f'TIMEOUT "{timeout}"' if len(timeout.strip()) > 0 else ''
-    if launcher[-3:] == ".sh":
-        dist_ut_port = process_dist_ut_port(2)
-        cmd += f'''if({archs} AND {os_})
-    bash_test_modules(
-    {name}
-    START_BASH
-    {launcher}
-    LABELS
-    "RUN_TYPE={run_type}"
-    ENVS
-    "PADDLE_DIST_UT_PORT={dist_ut_port};{envs}")
-    set_tests_properties({name} PROPERTIES  {time_out_str} RUN_SERIAL {run_serial})
-endif()
-'''
-    else:
-        cmd += f'''if({archs} AND {os_})
-    py_test_modules(
-    {name}
-    MODULES
-    {name}
-    ENVS
-    "{envs}")
-    set_tests_properties({name} PROPERTIES {time_out_str} RUN_SERIAL {run_serial})
-endif()
-'''
-    for _ in conditions:
-        cmd += f"endif()\n"
-    return cmd
-
-
-PROCESSED_DIR = set()
-
-
-def gen_cmakelists(current_work_dir):
-    print("procfessing dir:", current_work_dir)
-    if current_work_dir == "":
-        current_work_dir = "."
-
-    contents = os.listdir(current_work_dir)
-    sub_dirs = []
-    for c in contents:
-        c_path = os.path.join(current_work_dir, c)
-        if c_path in PROCESSED_DIR:
+            assert len(self.NO_CMAKE_DIR_WARNING) == 0, err_msg
+
+
+class CMakeGenerator():
+
+    def __init__(self, root_dirs):
+        self.PROCESSED_DIR = set()
+        self.port_manager = DistUTPortManager()
+        self.root_dirs = root_dirs
+
+    def prepare_dist_ut_port(self):
```

Comment:
Add comments  why get parent instead of down.

Reply:
done